File: scripts/typography/transform.lua
--@Transform

--require:${PROJECT_REQUIRES_AVIUTL2}
--information:Transform@${SCRIPT_NAME} v${PROJECT_VERSION} by ${PROJECT_AUTHOR}
--label:${LABEL}

--group:Pivot Point,true
local pivot_x = 0.0 --track@pivot_x:Pivot::X,-100000,100000,0,0.01
local pivot_y = 0.0 --track@pivot_y:Pivot::Y,-100000,100000,0,0.01
local pivot_z = 0.0 --track@pivot_z:Pivot::Z,-100000,100000,0,0.01
--trackgroup@pivot_x,pivot_y,pivot_z:Group::Pivot
--group:Position,true
local position_x = 0.0 --track@position_x:Position::X,-100000,100000,0,0.01
local position_y = 0.0 --track@position_y:Position::Y,-100000,100000,0,0.01
local position_z = 0.0 --track@position_z:Position::Z,-100000,100000,0,0.01
--trackgroup@position_x,position_y,position_z:Group::Position
--group:Rotation,true
local rotation_w = 0.0 --track@rotation_w:Rotation::W,-3600,3600,0,0.01
local rotation_x = 0.0 --track@rotation_x:Rotation::X,-3600,3600,0,0.01
local rotation_y = 0.0 --track@rotation_y:Rotation::Y,-3600,3600,0,0.01
local rotation_z = 0.0 --track@rotation_z:Rotation::Z,-3600,3600,0,0.01
--#define EULER XYZ Euler=5,XZY Euler=7,YXZ Euler=11,YZX Euler=15,ZXY Euler=19,ZYX Euler=21
local rotation_mode = 21 --select@rotation_mode:Rotation::Mode=21,Quaternion=0,Axis Angle=1,${EULER}
--trackgroup@rotation_x,rotation_y,rotation_z:Group::Rotation
--group:Scale,true
local scale_x = 100.0 --track@scale_x:Scale::X,-10000,10000,100,0.01
local scale_y = 100.0 --track@scale_y:Scale::Y,-10000,10000,100,0.01
local scale_z = 100.0 --track@scale_z:Scale::Z,-10000,10000,100,0.01
--trackgroup@scale_x,scale_y,scale_z:Group::Scale
--group:Target,true
local target_local_space = true --checksection@target_local_space:Target::Local Space,true
local target_world_space = false --checksection@target_world_space:Target::World Space,false
--group:Additional Options,false
local influence = 100.0 --track@influence:Influence,0,100,100,0.01

do
    local _ = obj.setanchor("position_x,position_y,position_z", 0, "line", "xyz")

    --#include "utilities.lua"
    local utils = require("utilities")
    local lerp = utils.lerp

    local vector = obj.module("Vector@${PROJECT_NAME}")
    local rotate = vector.rotate

    influence = influence * 0.01
    scale_x = scale_x * 0.01
    scale_y = scale_y * 0.01
    scale_z = scale_z * 0.01

    if target_world_space then
        local v = {
            (obj.ox - pivot_x * influence) * (1.0 + (scale_x - 1.0) * influence),
            (obj.oy - pivot_y * influence) * (1.0 + (scale_y - 1.0) * influence),
            (obj.oz - pivot_z * influence) * (1.0 + (scale_z - 1.0) * influence),
        }

        v = rotate(influence, rotation_mode, rotation_w, rotation_x, rotation_y, rotation_z, v)

        obj.ox = v[1]
        obj.oy = v[2]
        obj.oz = v[3]
    end

    if target_local_space then
        local rx, ry, rz = rotate(influence, rotation_mode, rotation_w, rotation_x, rotation_y, rotation_z)

        obj.cx = obj.cx + pivot_x * influence
        obj.cy = obj.cy + pivot_y * influence
        obj.cz = obj.cz + pivot_z * influence
        obj.rx = obj.rx + rx
        obj.ry = obj.ry + ry
        obj.rz = obj.rz + rz
        obj.sx = lerp(obj.sx, obj.sx * scale_x, influence)
        obj.sy = lerp(obj.sy, obj.sy * scale_y, influence)
        obj.sz = lerp(obj.sz, obj.sz * scale_z, influence)
    end

    obj.ox = obj.ox + position_x * influence
    obj.oy = obj.oy + position_y * influence
    obj.oz = obj.oz + position_z * influence
end

File: plugins/filters/intern/transform/intern/transform.cpp
#include "../transform.hpp"

#include <cmath>
#include <cstdint>  // IWYU pragma: keep

#pragma warning(push)
#pragma warning(disable : 4201)  // 非標準の無名構造体 (filter2.h FILTER_ITEM_COLOR)
#include <filter2.h>
#pragma warning(pop)

#include <intern/vector/vector.hpp>

namespace {
namespace vector = flow::vector;

constinit LOG_HANDLE *logger = nullptr;

auto group_pivot = FILTER_ITEM_GROUP(L"Pivot Point", true);
auto pivot_x = FILTER_ITEM_TRACK(L"Pivot::X", 0.0, -100000.0, 100000.0, 0.01);
auto pivot_y = FILTER_ITEM_TRACK(L"Pivot::Y", 0.0, -100000.0, 100000.0, 0.01);
auto pivot_z = FILTER_ITEM_TRACK(L"Pivot::Z", 0.0, -100000.0, 100000.0, 0.01);
FILTER_ITEM_TRACK *pivot[] = {&pivot_x, &pivot_y, &pivot_z, nullptr};
auto trackgroup_pivot = FILTER_ITEM_TRACK_GROUP(L"Group::Pivot", pivot);
auto group_position = FILTER_ITEM_GROUP(L"Position", true);
auto position_x = FILTER_ITEM_TRACK(L"Position::X", 0.0, -100000.0, 100000.0, 0.01);
auto position_y = FILTER_ITEM_TRACK(L"Position::Y", 0.0, -100000.0, 100000.0, 0.01);
auto position_z = FILTER_ITEM_TRACK(L"Position::Z", 0.0, -100000.0, 100000.0, 0.01);
FILTER_ITEM_TRACK *position[] = {&position_x, &position_y, &position_z, nullptr};
auto trackgroup_position = FILTER_ITEM_TRACK_GROUP(L"Group::Position", position);
auto group_rotation = FILTER_ITEM_GROUP(L"Rotation", true);
auto rotation_w = FILTER_ITEM_TRACK(L"Rotation::W", 0.0, -3600.0, 3600.0, 0.01);
auto rotation_x = FILTER_ITEM_TRACK(L"Rotation::X", 0.0, -3600.0, 3600.0, 0.01);
auto rotation_y = FILTER_ITEM_TRACK(L"Rotation::Y", 0.0, -3600.0, 3600.0, 0.01);
auto rotation_z = FILTER_ITEM_TRACK(L"Rotation::Z", 0.0, -3600.0, 3600.0, 0.01);
FILTER_ITEM_SELECT::ITEM rotation_modes[] = {
        {L"Quaternion", 0},
        {L"Axis Angle", 1},
        {L"XYZ Euler", 5},
        {L"XZY Euler", 7},
        {L"YXZ Euler", 11},
        {L"YZX Euler", 15},
        {L"ZXY Euler", 19},
        {L"ZYX Euler", 21},
        {nullptr, 0},
};
auto rotation_mode = FILTER_ITEM_SELECT(L"Rotation::Mode", 21, rotation_modes);
FILTER_ITEM_TRACK *rotation[] = {&rotation_x, &rotation_y, &rotation_z, nullptr};
auto trackgroup_rotation = FILTER_ITEM_TRACK_GROUP(L"Group::Rotation", rotation);
auto group_scale = FILTER_ITEM_GROUP(L"Scale", true);
auto scale_x = FILTER_ITEM_TRACK(L"Scale::X", 100.0, -10000.0, 10000.0, 0.01);
auto scale_y = FILTER_ITEM_TRACK(L"Scale::Y", 100.0, -10000.0, 10000.0, 0.01);
auto scale_z = FILTER_ITEM_TRACK(L"Scale::Z", 100.0, -10000.0, 10000.0, 0.01);
FILTER_ITEM_TRACK *scale[] = {&scale_x, &scale_y, &scale_z, nullptr};
auto trackgroup_scale = FILTER_ITEM_TRACK_GROUP(L"Group::Scale", scale);
auto group_target = FILTER_ITEM_GROUP(L"Target", true);
auto target_local_space = FILTER_ITEM_CHECK_SECTION(L"Target::Local Space", true);
auto target_world_space = FILTER_ITEM_CHECK_SECTION(L"Target::World Space", false);
auto group_additional_options = FILTER_ITEM_GROUP(L"Additional Options", false);
auto influence = FILTER_ITEM_TRACK(L"Influence", 100.0, 0.0, 100.0, 0.01);

void *props[] = {
        &group_pivot,
        &pivot_x,
        &pivot_y,
        &pivot_z,
        &trackgroup_pivot,
        &group_position,
        &position_x,
        &position_y,
        &position_z,
        &trackgroup_position,
        &group_rotation,
        &rotation_w,
        &rotation_x,
        &rotation_y,
        &rotation_z,
        &rotation_mode,
        &trackgroup_rotation,
        &group_scale,
        &scale_x,
        &scale_y,
        &scale_z,
        &trackgroup_scale,
        &group_target,
        &target_local_space,
        &target_world_space,
        &group_additional_options,
        &influence,
        nullptr,
};

constexpr bool
transform(FILTER_PROC_VIDEO *video) {
    const double t = influence.value * 0.01;
    const double sx = scale_x.value * 0.01, sy = scale_y.value * 0.01, sz = scale_z.value * 0.01;
    double rw = rotation_w.value, rx = rotation_x.value, ry = rotation_y.value, rz = rotation_z.value;

    if (rotation_mode.value == 1)
        rw = vector::to_rad(rw);
    else if (rotation_mode.value >= 5 && rotation_mode.value <= 21)
        rx = vector::to_rad(rx), ry = vector::to_rad(ry), rz = vector::to_rad(rz);

    if (target_world_space.value) {
        const Eigen::Vector3d v(
                (video->param->x - pivot_x.value * t) * (1.0 + (sx - 1.0) * t),
                (video->param->y - pivot_y.value * t) * (1.0 + (sy - 1.0) * t),
                (video->param->z - pivot_z.value * t) * (1.0 + (sz - 1.0) * t));

        std::vector<Eigen::Vector3d> vectors{v};

        vector::rotate(t, rotation_mode.value, rw, rx, ry, rz, vectors);

        video->param->x = static_cast<float>(vectors[0].x());
        video->param->y = static_cast<float>(vectors[0].y());
        video->param->z = static_cast<float>(vectors[0].z());
    }

    if (target_local_space.value) {
        const auto angle = vector::to_euler(t, rotation_mode.value, rw, rx, ry, rz);

        video->param->cx += static_cast<float>(pivot_x.value * t);
        video->param->cy += static_cast<float>(pivot_y.value * t);
        video->param->cz += static_cast<float>(pivot_z.value * t);

        video->param->rx += static_cast<float>(vector::to_deg(angle[0]));
        video->param->ry += static_cast<float>(vector::to_deg(angle[1]));
        video->param->rz += static_cast<float>(vector::to_deg(angle[2]));

        video->param->sx = static_cast<float>(std::lerp(video->param->sx, video->param->sx * sx, t));
        video->param->sy = static_cast<float>(std::lerp(video->param->sy, video->param->sy * sy, t));
        video->param->sz = static_cast<float>(std::lerp(video->param->sz, video->param->sz * sz, t));
    }

    if (target_local_space.value || target_world_space.value) {
        video->param->x += static_cast<float>(position_x.value * t);
        video->param->y += static_cast<float>(position_y.value * t);
        video->param->z += static_cast<float>(position_z.value * t);
    }

    return true;
}

constinit FILTER_PLUGIN_TABLE info = {
        .flag = FILTER_PLUGIN_TABLE::FLAG_VIDEO,
        .name = L"Transform@FlowType_K",
        .label = L"Text",
        .information = L"Transform@FlowType_K v" VERSION L" by Korarei",
        .items = props,
        .func_proc_video = transform,
        .func_proc_audio = nullptr,
};
}  // namespace

namespace flow::filter::transform {
void
init(HOST_APP_TABLE *host, LOG_HANDLE *handle) {
    logger = handle;

    host->register_filter_plugin(&info);
}
}  // namespace flow::filter::transform
